docs/plugins/testing_workshop_docs/Article-3-readers-and-fixtures.md

```diff
@@ -0,0 +1,152 @@
+ 
+        def write_func(filename):  
+            my_test_file = str(tmp_path / filename)  
+            original_data = np.random.rand(20, 20)  
+            np.save(my_test_file, original_data)  
+       
+            return my_test_file, original_data  
+ 
+        return write_func  
+   
+    def test_reader_round_trip(write_im_to_file):  
+        my_test_file, original_data = write_im_to_file("myfile.npy")  
+     
+        reader = napari_get_reader(my_test_file)  
+        assert callable(reader)  
+     
+        layer_data_list = reader(my_test_file)  
+        assert isinstance(layer_data_list, List) and len(layer_data_list) > 0  
+     
+        layer_data_tuple = layer_data_list[0]  
+        layer_data = layer_data_tuple[0]  
+        np.testing.assert_allclose(layer_data, original_data)  
+``` 
+We’re going to assert a list length greater than zero. There must be a layer in there; otherwise, we didn't read it correctly. We also assert that it is a list. Once we start testing, there are many things we could test. Try to focus on what's most useful.  
+  
+We will test that inside that list is what we expected. Remember, it's a list of layer data tuples. The first item of a tuple of the layer data tuple is the actual data. We’re going to test that explicitly.  
+  
+Then we assert, using `numpy`’s asserting mechanism, `np.testing.assert_allclose` that they are all close, even though they should be exactly the same. This is standard practice when working with floating point precision. Numpy also has [other assertion options](https://numpy.org/doc/stable/reference/routines.testing.html) you may find useful. The layer data we read back with the reader function should be the same as the original data. If that's true, then we made the entire round trip. We saved the file and we used the reader to read the file.  
+```python
+ def test_reader_round_trip(write_im_to_file):  
+        my_test_file, original_data = write_im_to_file("myfile.npy")  
+     
+        reader = napari_get_reader(my_test_file)  
+        assert callable(reader)  
+     
+        layer_data_list = reader(my_test_file)  
+        assert isinstance(layer_data_list, List) and len(layer_data_list) > 0  
+     
+        layer_data_tuple = layer_data_list[0]  
+        layer_data = layer_data_tuple[0]  
+        np.testing.assert_allclose(layer_data, original_data)  
+```    
+We run our tests again, and now two are collected, both passing.  
+
+![pytest - tests passed](../../images/napari_plugins_5th_tests_passed.png)
+
+  
+## Enclosed testing  
+We did not need a viewer or napari to test this. It's important that we didn't need those because napari and the napari viewer are out of our control. What we can control is the code the _we_ wrote. We wrote that data by simply mocking up some data and getting a temporary path to it. We could thoroughly test our functions in an enclosed way without relying on other people's code or mocking up many complicated objects.  
```

Comment:
```suggestion
Note that although we're testing a `napari` plugin, we did not need a viewer or napari to test this. It's important that we didn't need those because napari and the napari viewer are out of our control. What we can control is the code _we_ wrote. We wrote that data by simply mocking up an array and getting a temporary path to it. We could thoroughly test our functions in an enclosed way without relying on other people's code or mocking up many complicated objects.  
```